src/context/TodoContext.tsx

```diff
@@ -0,0 +1,149 @@
+import React, {
+  MutableRefObject,
+  ReactNode,
+  SetStateAction,
+  createContext, useContext, useEffect, useRef, useState,
+} from 'react';
+import { editTodo, getTodos } from '../api/todos';
+import { ErrorTypes, SortTypes, Todo } from '../types/Todo';
+
+export interface TContext {
+  todos: Todo[];
+  setTodos: React.Dispatch<SetStateAction<Todo[]>>;
+  hasError: null | ErrorTypes;
+  handleError: (error: ErrorTypes) => void;
+  sortType: SortTypes;
+  setSortType: React.Dispatch<SetStateAction<SortTypes>>;
+  setHasError: React.Dispatch<SetStateAction<ErrorTypes | null>>;
+  tempTodos: Todo | null;
+  setTempTodos: React.Dispatch<SetStateAction<Todo | null>>;
+  handleToggleStatus: (todoId: number) => void ;
+  isToggled: boolean,
+  setIsToggled: React.Dispatch<SetStateAction<boolean>>,
+  handleToggleAllStatus:() => void,
+  isToggledAll: boolean,
+  setIsToggledAll: React.Dispatch<SetStateAction<boolean>>,
+  titleInputRef: MutableRefObject<HTMLInputElement | null>,
+  editedRef: MutableRefObject<HTMLInputElement | null>,
+  isGroupDeleting: boolean,
+  setIsGroupDeleting: React.Dispatch<SetStateAction<boolean>>,
+}
+
+const TodoContext = createContext<TContext | null>(null);
+
+export function useTodoContext() {
+  return useContext(TodoContext);
+}
+
+const USER_ID = 11550;
+
+export function TodoProvider({ children }: { children: ReactNode }) {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [hasError, setHasError] = useState<ErrorTypes | null>(null);
+  const [sortType, setSortType] = useState<SortTypes>('all');
+  const [tempTodos, setTempTodos] = useState<Todo | null>(null);
+  const [isToggled, setIsToggled] = useState<boolean>(false);
+  const [isToggledAll, setIsToggledAll] = useState<boolean>(false);
+  const titleInputRef = useRef<HTMLInputElement | null>(null);
+  const [isGroupDeleting, setIsGroupDeleting] = useState<boolean>(false);
+  const editedRef = useRef<HTMLInputElement | null>(null);
+
+  const handleError = (error: ErrorTypes) => {
+    setHasError(error);
+    setTimeout(() => setHasError(null), 3000);
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then((res) => {
+        setTodos(res);
+      })
+      .catch(() => {
+        handleError('Unable to load todos');
+      });
+  }, []);
+
+  const handleToggleStatus = (todoId: number) => {
+    const updatedTodoIndex = todos
+      .findIndex((toggledtodo) => toggledtodo.id === todoId);
+
+    if (updatedTodoIndex !== -1) {
+      const updatedTodo = todos[updatedTodoIndex];
+      const updatedTodoWithNewStatus
+      = { ...updatedTodo, completed: !updatedTodo.completed };
+
+      setIsToggled(true);
+
+      editTodo(todoId, updatedTodoWithNewStatus)
```

Comment:
You don't have to pass the whole `Todo` to the `PATCH`.
It's always good to transfer at least amount of data between backend and frontend.

Here we are only interested to set the completed status to the opposite, so the payload should be 
```
{
    completed: !updatedTodo.completed
}
```

Please refer to the API documentation for further information: https://mate-academy.github.io/fe-students-api/todos.html